data/transactions/logic/frames.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opProto(cx *EvalContext) error {
+	if !cx.fromCallsub {
+		return fmt.Errorf("proto executed in normal flow")
+	}
+	cx.fromCallsub = false
+	nargs := int(cx.program[cx.pc+1])
+	if nargs > len(cx.stack) {
+		return fmt.Errorf("callsub to proto that requires %d args with stack height %d", nargs, len(cx.stack))
+	}
+	top := len(cx.callstack) - 1
+	cx.callstack[top].clear = true
+	cx.callstack[top].args = nargs
+	cx.callstack[top].returns = int(cx.program[cx.pc+2])
+	return nil
+}
+
+func opFrameDig(cx *EvalContext) error {
+	i := int8(cx.program[cx.pc+1])
+
+	top := len(cx.callstack) - 1
+	if top < 0 {
+		return errors.New("frame_dig with empty callstack")
+	}
+
+	frame := cx.callstack[top]
+	if i < 0 && -int(i) > frame.args {
+		return fmt.Errorf("frame_dig %d in sub with %d args", i, frame.args)
+	}
+	idx := frame.height + int(i)
+	if idx >= len(cx.stack) {
+		return errors.New("frame_dig above stack")
+	}
+	if idx < 0 { // Impossible, because of arglen check in callsub
```

Comment:
The impossibility comes from 
```
	nargs := int(cx.program[cx.pc+1])
	if nargs > len(cx.stack) {
		return fmt.Errorf("callsub to proto that requires %d args with stack height %d", nargs, len(cx.stack))
	}
```
This ensures that when calling a `proto` subroutine, you have supplied enough args.  So in this code, the stack "must" be big enough for you.

Of course, I now see that's not true.  You might (erroneously, I presume) pop your args off the stack, and then use `frame_dig -3` anyway.  The stack was big enough on entry, but you shrank it before using `frame_dig`.  I will remove the comment.

Reply:
Well, no, it does remain impossible - since you are subtracting from the frame.height, and you have that previous checks that the frame.height was higher than len(frame.args).

However, it *won't* remain impossible if I remove that check on non-proto'd subroutines, which I think I should do.

Reply:
I think this is reolved - it is now _certainly_ possible, and tests have been added to show it works.